### ISSUE TITLE
Better error message for stack overflows on Unix

### ISSUE THREAD

**Issue by stephentoub** [dotnet/coreclr#1504]
Consider this program:

``` C#
class Program
{
    static unsafe void Main()
    {
        for (int i = 0; ; i++)
        {
            byte* bytes = stackalloc byte[1024];
            System.Console.WriteLine(i);
        }
    }
}
```

On Windows, this stack overflows and prints out a message:

```
Process is terminated due to StackOverflowException.
```

On Unix, it simply seg faults with little indication as to why.  If it's feasible, it'd be very helpful if some kind of error message or indication could be generated to help the developer understand quickly why things seg faulted.  If nothing else, some analytics in SOS that could point the finger at a stack overflow would be helpful.

cc: @jkotas, @janvorli, @mikem8361

**Comment by sergiy-k** [comment]
@janvorli, is there anything we can do in addition to https://github.com/dotnet/coreclr/pull/2640?

**Comment by janvorli** [comment]
The only additional thing we could do would be to set the sigsegv handler run on a separate stack so that we can handle cases where we run out of stack in a tight recursive loop with only small size of locals where jitter doesn't add stack probes. But that would require having such a stack for each thread  (I guess one memory page plus one guard page might be enough) and also switching back to the main stack after we detect it was not a stack overflow (we don't want the regular exception handling to run on the alternate stack).
I think it is worth trying to do that.

**Comment by davjohna** [comment]
1. Is the change still looking like it'll make it into 1.1.0?
2. Will this change fix the error message for all the ways a StackOverflowException could occur, or just a small subset of them?

**Comment by janvorli** [comment]
@davjohna - it is not planned for 1.1.0. We already generate error message for StackOverflowException in cases where we can detect it before we actually get out of the stack. These are cases where jit inserts explicit checks for stack space. It happens when a function has locals larger than 4kB. 
So this remaining case is when there are no explicit checks and we physically run out of stack.

**Comment by janvorli** [comment]
I'm working on a solution based on alternate stack for SIGSEGV.

**Comment by janvorli** [comment]
I have a prototype working on amd64 Linux. I need to add support for other architectures, but it looks promising.

**Comment by janvorli** [comment]
Fixed by #9650
